{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "269c0de9-dd10-45b0-b596-fb2ea183e720",
   "metadata": {},
   "source": [
    "# Murphy Angelo & Cyrus Tau project report\n",
    "\n",
    "### Detecting riboswitches with diffusion + SOFM\n",
    "\n",
    "RNA flexibility is a key challenge in predicting RNA structures and is the defining characteristic of riboswitches, which show promise as selective gene regulation platforms. We used a diffusion model to generate 5000 structures for 4053 riboswitch sequences and 4053 shuffled controls. For each of the ~8000 sequences, we applied SOFMs to analyze the 5000 structures (represented as LxL binary tensors) to quantify alternative conformations.  We used a simple logistic regressor trained on these features to achieve state-of-the-art riboswitch detection vs. established methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ad6b6488-c3bd-4a3c-8de5-2acecb1ff4a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import matplotlib\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "from joblib import Parallel, delayed\n",
    "from karateclub import Graph2Vec\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import ct_tools as ct\n",
    "ct.setup_figure_saving('murfalo_cyrustau_project')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f16e0926-24b7-41f9-8bac-d7a38db5aecf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_x0(prediction: Path):\n",
    "    \"\"\"\n",
    "    Loads predictions as NxLxL matrices. N := number of diffusion samples.\n",
    "    L := sequence length.\n",
    "    \"\"\"\n",
    "    out_file = prediction / \"x0_compressed.npz\"\n",
    "    with np.load(out_file) as data:\n",
    "        x0, seq = data[\"x0\"], data[\"seq\"]\n",
    "    return seq, x0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ce7e10bc-0bf1-472e-94cd-27f5352a5cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "pred_dir = Path(\"./predict_results\")\n",
    "predictions = list(\n",
    "    path for path in pred_dir.iterdir() if not path.name.endswith(\"_shuffled\")\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9fba457-af11-448c-9bc9-d4dcfc04313c",
   "metadata": {},
   "source": [
    "### Prediction suffix naming rules\n",
    "- \"\" => normal sequence\n",
    "- \"_shuffled\" => mononucleotide shuffle\n",
    "- \"_shuffled_dinuc\" => dinucleotide shuffle (preserves dinuc. frequencies)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "00cdf234-ca1e-4c76-9c8a-9871d00ea157",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculate graph2vec embeddings of RNA adjacency matrices\n",
    "# NOTE(MCA): Very slow, better to run on HPC\n",
    "ignore_cache = False\n",
    "if ignore_cache == True:\n",
    "\n",
    "    def compute_embedding(prediction):\n",
    "        model = Graph2Vec(dimensions=128, workers=1)\n",
    "        _, x0 = load_x0(prediction)  # (N, L, L)\n",
    "        # x0 = (x0 + x0.transpose(0, 2, 1)) / 2.0  # enforce symmetry\n",
    "        graphs = [nx.from_numpy_array(a) for a in x0]\n",
    "\n",
    "        model.fit(graphs)\n",
    "        return model.get_embedding()  # (N, d)\n",
    "\n",
    "    embeddings = Parallel(n_jobs=-1, backend=\"loky\")(\n",
    "        delayed(compute_embedding)(p) for p in predictions\n",
    "    )\n",
    "    embeddings = np.stack(embeddings)  # (P, N, d)\n",
    "    for prediction, embedding in zip(predictions, embeddings):\n",
    "        np.savez_compressed(\n",
    "            prediction / \"g2v.npz\", embedding=embedding.astype(np.float32)\n",
    "        )\n",
    "\n",
    "\n",
    "def load_g2v(path: Path) -> np.ndarray:\n",
    "    return np.load(path / \"g2v.npz\")[\"embedding\"]\n",
    "\n",
    "\n",
    "# Load embeddings\n",
    "embeddings = np.stack(\n",
    "    Parallel(n_jobs=-1, backend=\"loky\")(delayed(load_g2v)(p) for p in predictions)\n",
    ")\n",
    "\n",
    "embeddings = np.stack([np.load(p / \"g2v.npz\")[\"embedding\"] for p in predictions])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "85dccb4c-2962-492c-b4cb-437c69ba822c",
   "metadata": {},
   "outputs": [],
   "source": [
    "from minisom import MiniSom\n",
    "\n",
    "\n",
    "def som(prediction, embedding) -> MiniSom:\n",
    "    # SOM hyperparameters\n",
    "    m_neurons = 2\n",
    "    n_neurons = 2\n",
    "    sigma = (m_neurons / n_neurons) / 2.0\n",
    "    eta = 0.5\n",
    "\n",
    "    # Generate SOMs for each RNA\n",
    "    N, L = embedding.shape\n",
    "    som = MiniSom(\n",
    "        m_neurons,\n",
    "        n_neurons,\n",
    "        L,\n",
    "        sigma=sigma,\n",
    "        learning_rate=eta,\n",
    "        neighborhood_function=\"gaussian\",\n",
    "        topology=\"rectangular\",\n",
    "        random_seed=42,\n",
    "    )\n",
    "    som.pca_weights_init(embedding)\n",
    "    som.train(embs, N, verbose=False)\n",
    "    return som\n",
    "\n",
    "\n",
    "soms = list(\n",
    "    Parallel(n_jobs=-1, backend=\"loky\")(\n",
    "        delayed(som)(p, emb) for p, emb in zip(predictions, embeddings)\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "ffde2c5a-ad36-40ca-9258-35b7bd557757",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/5d/hx6lx2b96_q3ty9csdybv1440000gn/T/ipykernel_12569/1198385565.py:20: UserWarning: No artists with labels found to put in legend.  Note that artists whose label start with an underscore are ignored when legend() is called with no argument.\n",
      "  plt.legend(loc='upper right')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 1000x900 with 2 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plot SOM U matrices\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "%matplotlib inline\n",
    "\n",
    "som, embs = soms[1], embeddings[1]\n",
    "w_x, w_y = zip(*[som.winner(d) for d in embs])\n",
    "w_x = np.array(w_x)\n",
    "w_y = np.array(w_y)\n",
    "\n",
    "plt.figure(figsize=(10, 9))\n",
    "plt.pcolor(som.distance_map().T, cmap=\"bone_r\", alpha=0.2)\n",
    "plt.colorbar()\n",
    "plt.scatter(\n",
    "    w_x + 0.5 + (np.random.rand(len(w_x)) - 0.5) * 0.8,\n",
    "    w_y + 0.5 + (np.random.rand(len(w_y)) - 0.5) * 0.8,\n",
    "    s=50,\n",
    "    c=\"C0\",\n",
    ")\n",
    "\n",
    "plt.legend(loc=\"upper right\")\n",
    "plt.grid()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "a962a901-2b1a-49e5-bd1d-28619991908e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/Caskroom/miniforge/base/envs/9530/lib/python3.10/site-packages/joblib/externals/loky/process_executor.py:782: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Create feature vector and labels for training\n",
    "\n",
    "\n",
    "def get_features(prediction, som, embs):\n",
    "    \"\"\"\n",
    "    Returns X, y (features and labels) for a given prediction, som, and embeddings.\n",
    "    \"\"\"\n",
    "    _, x0 = load_x0(prediction)  # (N, L, L)\n",
    "    p_bp = x0.max(axis=1).mean(axis=(0, 1))  # avg. base pair probability\n",
    "    qe = som.quantization_error(embs)\n",
    "    return (x0_frob, p_bp, qe, prediction.name.endswith(\"_shuffled_dinuc\"))\n",
    "\n",
    "\n",
    "features = np.stack(\n",
    "    Parallel(n_jobs=-1, backend=\"loky\")(\n",
    "        delayed(get_features)(p, som, emb)\n",
    "        for p, som, emb in zip(predictions, soms, embeddings)\n",
    "    )\n",
    ")\n",
    "X, y = features[:, :-1], features[:, -1]\n",
    "del features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "eb7a660c-e9e1-4124-8f3c-baf7289b6d63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicting AdoCbl...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/homebrew/Caskroom/miniforge/base/envs/9530/lib/python3.10/site-packages/joblib/externals/loky/process_executor.py:782: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Predicting Cobalamin...\n",
      "Predicting FMN...\n",
      "Predicting Fluoride...\n",
      "Predicting Glycine...\n",
      "Predicting Lysine...\n",
      "Predicting MFR...\n",
      "Predicting MOCO...\n",
      "Predicting NiCo...\n",
      "Predicting PreQ1...\n",
      "Predicting PreQ1-III...\n",
      "Predicting Purine...\n",
      "Predicting SAH...\n",
      "Predicting SAM...\n",
      "Predicting SAM-I-IV-variant...\n",
      "Predicting SAM-IV...\n",
      "Predicting SAM-SAH...\n",
      "Predicting SMK...\n",
      "Predicting THF...\n",
      "Predicting TPP...\n",
      "Predicting ZMP-ZTP...\n",
      "Predicting c-di-GMP-I...\n",
      "Predicting c-di-GMP-II...\n",
      "Predicting glnA...\n",
      "Predicting preQ1-II...\n",
      "Cross-validated F1: 0.78283041\n",
      "Cross-validated AUC: 0.77944846\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.metrics import (\n",
    "    precision_recall_curve,\n",
    "    roc_auc_score,\n",
    "    balanced_accuracy_score,\n",
    "    matthews_corrcoef,\n",
    ")\n",
    "\n",
    "# Simple logistic regressor with LOOCV across families\n",
    "accessions, families = zip(*[p.name.split(\"_\")[0:2] for p in predictions])\n",
    "accessions, families = np.stack(accessions), np.stack(families)\n",
    "family_to_accession = {\n",
    "    family: accession for family, accession in zip(families, accessions)\n",
    "}\n",
    "\n",
    "unique_families = sorted(set(families))\n",
    "family_metrics = []\n",
    "for family in unique_families:\n",
    "    print(f\"Predicting {family}...\")\n",
    "    # Split data into training and test sets\n",
    "    test_mask = families == family\n",
    "    train_mask = ~test_mask\n",
    "    X_train = X[train_mask]\n",
    "    y_train = y[train_mask]\n",
    "    X_test = X[test_mask]\n",
    "    y_test = y[test_mask]\n",
    "\n",
    "    # Skip empty families (only possible if all runs are not finished)\n",
    "    if len(np.unique(y_test)) < 2:\n",
    "        continue\n",
    "\n",
    "    # Standardize features: fit scaler on training data only to avoid leakage\n",
    "    scaler = StandardScaler()\n",
    "    X_train_scaled = scaler.fit_transform(X_train)\n",
    "    X_test_scaled = scaler.fit_transform(X_test)\n",
    "\n",
    "    # Hyperparameter search\n",
    "    param_grid = {\n",
    "        \"C\": [0.01, 0.1, 1, 10],\n",
    "        \"penalty\": [\"elasticnet\"],\n",
    "        \"solver\": [\"saga\"],\n",
    "        \"l1_ratio\": [0.0, 0.2, 0.5, 0.8, 1.0],  # 0=L2, 1=L1\n",
    "    }\n",
    "    clf = GridSearchCV(\n",
    "        LogisticRegression(max_iter=10000),\n",
    "        param_grid,\n",
    "        cv=5,\n",
    "        scoring=\"roc_auc\",\n",
    "        n_jobs=-1,\n",
    "    )\n",
    "    clf.fit(X_train_scaled, y_train)\n",
    "    best_clf = clf.best_estimator_\n",
    "\n",
    "    # Predictions and scores\n",
    "    y_scores_test = best_clf.predict_proba(X_test_scaled)[:, 1]\n",
    "    y_pred_test = best_clf.predict(X_test_scaled)\n",
    "\n",
    "    # Compute and append metrics\n",
    "    with np.errstate(divide=\"ignore\", invalid=\"ignore\"):\n",
    "        precision, recall, _ = precision_recall_curve(y_test, y_scores_test)\n",
    "        F1_scores = np.nan_to_num(2 * (precision * recall) / (precision + recall))\n",
    "        max_f1 = np.max(F1_scores)\n",
    "        auc = roc_auc_score(y_test, y_scores_test)\n",
    "        balanced_acc = balanced_accuracy_score(y_test, y_pred_test)\n",
    "        mcc = matthews_corrcoef(y_test, y_pred_test)\n",
    "\n",
    "        # Compute AUC\n",
    "        family_metrics.append(\n",
    "            {\n",
    "                \"accession\": family_to_accession[family],\n",
    "                \"family\": family,\n",
    "                \"num_samples\": np.sum(test_mask),\n",
    "                \"max_f1\": max_f1,\n",
    "                \"auc\": auc,\n",
    "                \"balanced_acc\": balanced_acc,\n",
    "                \"mcc\": mcc,\n",
    "                \"avg_length\": np.mean(X[test_mask, -1]),\n",
    "            }\n",
    "        )\n",
    "\n",
    "# Compute and print average metrics across all families\n",
    "avg_max_f1 = np.mean([m[\"max_f1\"] for m in family_metrics])\n",
    "avg_auc = np.mean([m[\"auc\"] for m in family_metrics])\n",
    "print(f\"Cross-validated F1: {avg_max_f1:.8f}\")\n",
    "print(f\"Cross-validated AUC: {avg_auc:.8f}\")\n",
    "\n",
    "# Interpretable decision boundary on all unscaled data\n",
    "if X.shape[1] == 2 and len(np.unique(y)) > 1:\n",
    "    param_grid = {\n",
    "        \"C\": [0.01, 0.1, 1, 10],\n",
    "        \"penalty\": [\"elasticnet\"],\n",
    "        \"solver\": [\"saga\"],\n",
    "        \"l1_ratio\": [0.0, 0.2, 0.5, 0.8, 1.0],  # 0=L2, 1=L1\n",
    "    }\n",
    "    final_clf_search = GridSearchCV(\n",
    "        LogisticRegression(max_iter=10000),\n",
    "        param_grid,\n",
    "        cv=5,\n",
    "        scoring=\"roc_auc\",\n",
    "        n_jobs=-1,\n",
    "    )\n",
    "    final_clf_search.fit(X, y)\n",
    "    final_clf = final_clf_search.best_estimator_\n",
    "    # FIXME(MCA): Restore feature labels\n",
    "    # plot_fig(\n",
    "    #     X,\n",
    "    #     y,\n",
    "    #     [\"Mean Contact Strength\", \"Ensemble Concentration\"],\n",
    "    #     coef=final_clf.coef_,\n",
    "    #     intercept=final_clf.intercept_,\n",
    "    # )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "ab9e2588-5822-49d0-936d-cc325db7565a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Figure size 800x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Compare Goodarzi performance vs. ours\n",
    "# --- Plot styling ---\n",
    "plt.rcParams[\"font.size\"] = 16\n",
    "plt.rcParams[\"axes.titlesize\"] = 24\n",
    "plt.rcParams[\"axes.labelsize\"] = 18\n",
    "plt.rcParams[\"xtick.labelsize\"] = 16\n",
    "plt.rcParams[\"ytick.labelsize\"] = 16\n",
    "plt.rcParams[\"legend.fontsize\"] = 16\n",
    "\n",
    "# --- Load data ---\n",
    "df_mine = pd.DataFrame(family_metrics)\n",
    "df_goodarzi = pd.read_csv(\"goodarzi_metrics.csv\")\n",
    "\n",
    "# --- Process data ---\n",
    "overlapping_families = list(\n",
    "    set(df_mine[\"accession\"]) & set(df_goodarzi[\"Rfam family identifier\"])\n",
    ")\n",
    "overlapping_families.sort()\n",
    "auc_mine = [\n",
    "    df_mine[df_mine[\"accession\"] == family][\"auc\"].values[0]\n",
    "    for family in overlapping_families\n",
    "]\n",
    "auc_goodarzi = [\n",
    "    df_goodarzi[df_goodarzi[\"Rfam family identifier\"] == family][\"AUC value\"].values[0]\n",
    "    for family in overlapping_families\n",
    "]\n",
    "n_families = len(overlapping_families)\n",
    "\n",
    "# --- Plot the data ---\n",
    "fig, ax = plt.subplots(figsize=(8, 6))\n",
    "for i in range(n_families):\n",
    "    color = \"black\" if auc_mine[i] > auc_goodarzi[i] else \"#bf616a\"\n",
    "    ax.plot(\n",
    "        [0.4, 1.6],\n",
    "        [auc_goodarzi[i], auc_mine[i]],\n",
    "        color=color,\n",
    "        alpha=0.7,\n",
    "        linestyle=\"dotted\",\n",
    "        linewidth=1.5,\n",
    "        zorder=1,\n",
    "    )\n",
    "\n",
    "# Scatter\n",
    "ax.scatter(\n",
    "    [0.4] * n_families,\n",
    "    auc_goodarzi,\n",
    "    color=\"#8be9fd\",\n",
    "    s=100,\n",
    "    label=\"Goodarzi\",\n",
    "    alpha=0.9,\n",
    "    edgecolor=\"k\",\n",
    "    linewidths=2,\n",
    "    zorder=2,\n",
    ")\n",
    "ax.scatter(\n",
    "    [1.6] * n_families,\n",
    "    auc_mine,\n",
    "    color=\"#ffb86c\",\n",
    "    s=100,\n",
    "    label=\"Ours\",\n",
    "    alpha=0.9,\n",
    "    edgecolor=\"k\",\n",
    "    linewidths=2,\n",
    "    zorder=2,\n",
    ")\n",
    "\n",
    "# Axes\n",
    "ax.set_xticks([0.4, 1.6])\n",
    "ax.set_xticklabels([\"Goodarzi\", \"Ours\"], fontweight=\"bold\", fontsize=14)\n",
    "ax.set_xlim(0, 2)\n",
    "ax.set_ylabel(\"AUC\", fontweight=\"bold\", fontsize=14)\n",
    "ax.set_ylim(0, 1)\n",
    "\n",
    "# Title\n",
    "ax.set_title(\"Goodarzi vs. Ours\", fontweight=\"bold\", pad=25, fontsize=18)\n",
    "\n",
    "# Background\n",
    "ax.set_facecolor(\"#f5f5f5\")\n",
    "ax.grid(True, axis=\"y\", linestyle=\"--\", alpha=0.5, color=\"gray\")\n",
    "\n",
    "# Border\n",
    "for spine in ax.spines.values():\n",
    "    spine.set_edgecolor(\"black\")\n",
    "    spine.set_linewidth = 2\n",
    "\n",
    "# Layout + save\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"Versus Goodarzi.png\", dpi=300, bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bcaafce-b0b6-483f-a643-12fa138618f1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
